staging/src/github.com/kubeedge/beehive/pkg/core/module.go

```diff
@@ -18,27 +17,76 @@ type Module interface {
 
 var (
 	// Modules map
-	modules         map[string]Module
-	disabledModules map[string]Module
+	modules         map[string]moduleInfo
+	disabledModules map[string]moduleInfo
 )
 
 func init() {
-	modules = make(map[string]Module)
-	disabledModules = make(map[string]Module)
+	modules = make(map[string]moduleInfo)
+	disabledModules = make(map[string]moduleInfo)
+}
+
+// moduleInfo represent a module info
+type moduleInfo struct {
+	moduleType string
+	remote     bool
+	module     Module
 }
 
 // Register register module
-func Register(m Module) {
-	if m.Enable() {
-		modules[m.Name()] = m
-		klog.Infof("Module %v registered successfully", m.Name())
+// if not passed in parameter opts, default moduleType is "channel"
+func Register(m Module, opts ...string) {
+	var info moduleInfo
+	if len(opts) == 0 {
+		info = moduleInfo{
+			module:     m,
+			moduleType: common.MsgCtxTypeChannel,
+			remote:     false,
+		}
 	} else {
-		disabledModules[m.Name()] = m
-		klog.Warningf("Module %v is disabled, do not register", m.Name())
+		info = moduleInfo{
+			module:     m,
+			moduleType: opts[0],
+			remote:     true,
+		}
+	}
+	coreContext := context.GetGlobalContext()
+	if IsModuleEnabled(m) {
+		//Init the module
+		add := context.AddModuleInfo{
+			ModuleName: info.module.Name(),
+			ModuleType: info.moduleType,
+			ModuleSocket: context.ModuleSocket{
+				IsRemote: info.remote,
+			},
+		}
+		coreContext.AddModule(&add)
+		//Assemble typeChannels for sendToGroup
+		coreContext.AddModuleGroup(m.Name(), info.module.Group())
+		modules[m.Name()] = info
+		klog.Infof("module %s registered", m.Name())
+	} else {
+		disabledModules[m.Name()] = info
+		klog.Infof("module %s is not register", m.Name())
 	}
 }
 
+// IsModuleEnabled return the module status
+func IsModuleEnabled(m Module) bool {
+	return m.Enable()
+}
+
 // GetModules gets modules map
-func GetModules() map[string]Module {
+func GetModules() map[string]moduleInfo {
 	return modules
 }
+
+// Cleanup modules
+func Cleanup() {
+	return
+}
+
+// GetModule gets module
+func (m *moduleInfo) GetModule() Module {
```

Comment:
Who will call this func? If not, please delete

Reply:
by metamanger_test.go, need to get the field `module`, which is not exported